File: Assets/_Project/Scripts/Gameplay/Weapons/OrbitWeapon.cs
using UnityEngine;
using VampireSurvivors.Core;

namespace VampireSurvivors.Gameplay
{
    /// <summary>
    /// 環繞武器（如聖經） — 投射物環繞角色旋轉，持續對接觸的敵人造成傷害。
    /// 投射物從 Object Pool 取出，環繞一定時間後回收。
    /// </summary>
    public class OrbitWeapon : BaseWeapon
    {
        /// <summary>環繞旋轉速度（度/秒）。</summary>
        private const float OrbitSpeed = 180f;

        /// <summary>環繞持續時間（秒）。</summary>
        private const float OrbitDuration = 3f;

        /// <summary>
        /// 建構環繞武器。
        /// </summary>
        /// <param name="config">武器設定檔。</param>
        /// <param name="poolManager">物件池管理器。</param>
        public OrbitWeapon(WeaponConfigSO config, ObjectPoolManager poolManager)
            : base(config, poolManager) { }

        /// <inheritdoc/>
        protected override void OnAttack(Vector2 origin, Vector2 direction)
        {
            var data = CurrentLevelData;
            float angleStep = 360f / data.projectileCount;

            for (int i = 0; i < data.projectileCount; i++)
            {
                float angle = i * angleStep;
                Vector2 offset = new Vector2(
                    Mathf.Cos(angle * Mathf.Deg2Rad),
                    Mathf.Sin(angle * Mathf.Deg2Rad)
                ) * data.attackRange;

                var obj = poolManager.Spawn(config.weaponId, origin + offset);
                var orbit = obj.GetComponent<OrbitProjectile>();
                if (orbit != null)
                {
                    orbit.Initialize(origin, data.attackRange, OrbitSpeed, angle,
                        data.damage, OrbitDuration, config.weaponId, poolManager);
                }
            }
        }
    }
}

File: src/gameplay/weapons/weapons/OrbitWeapon.ts
import Phaser from 'phaser';
import { WeaponConfig } from '../../../data/types';
import { BaseWeapon } from './BaseWeapon';
import { ObjectPoolManager } from '../../../core/pool/ObjectPoolManager';

/**
 * 環繞武器（太極環、烈焰掌）
 * 在玩家周圍生成環繞的投射物，持續傷害接觸的敵人。
 * projectileCount 決定環繞數量，attackRange 決定環繞半徑。
 */
export class OrbitWeapon extends BaseWeapon {
  private poolManager: ObjectPoolManager;

  constructor(scene: Phaser.Scene, config: WeaponConfig, poolManager: ObjectPoolManager) {
    super(scene, config);
    this.poolManager = poolManager;
  }

  attack(origin: Phaser.Math.Vector2, _direction: Phaser.Math.Vector2): void {
    const ld = this.getLevelData();
    const range = ld.attackRange;
    const count = Math.max(1, ld.projectileCount);

    try {
      const pool = this.poolManager.getPool('projectile');

      // 生成環繞投射物，均勻分布在圓周上
      for (let i = 0; i < count; i++) {
        const angle = (Math.PI * 2 / count) * i + (Date.now() * 0.003); // 旋轉動畫
        const x = origin.x + Math.cos(angle) * range;
        const y = origin.y + Math.sin(angle) * range;

        // AoE 投射物：靜止、短生命週期、有 aoeRadius 可穿透打多敵
        pool.spawn(x, y, ld.damage, 0, 0, 0.35, range * 0.4);
      }
    } catch {
      // Pool not registered
    }
  }
}
